src/librustc_trans/context.rs

```diff
@@ -843,6 +843,10 @@ impl<'a, 'tcx> LayoutTyper<'tcx> for &'a SharedCrateContext<'a, 'tcx> {
     type TyLayout = TyLayout<'tcx>;
 
     fn layout_of(self, ty: Ty<'tcx>) -> Self::TyLayout {
+        if let Some(&layout) = self.tcx().layout_cache.borrow().get(&ty) {
+            return TyLayout { ty: ty, layout: layout, variant_index: None };
+        }
+
         self.tcx().infer_ctxt((), traits::Reveal::All).enter(|infcx| {
```

Comment:
Is `has_projection_types` anything other than a flag check? I'm not sure I understand what's going on. Can the cache be hit with the unnormalized type if `has_projection_types` returns true?